src/lex.cpp

```diff
@@ -1288,6 +1292,52 @@ void lexer::parse_modern_octal_number() {
       diag_unexpected_characters_in_octal_number>(input);
 }
 
+void lexer::check_precision_loss(const char8* number_begin,
+                                 const char8* input) {
```

Comment:
Nit: Rename `input` to `number_end`. `input` doesn't make sense.

Better yet, accept a string8_view instead:
```suggestion
void lexer::check_precision_loss(string8_view number) {
```

Reply:
Right now, I'm setting `characters` to `source_code_span(number_begin, input)` when I emit my error message. If I change the function to accept a string8_view, how can I use that to create a source_code_span?

Reply:
> If I change the function to accept a string8_view, how can I use that to create a source_code_span?

`source_code_span(number.data(), number.data() + number.size())`

To make this easier, we should probably add a `source_code_span` constructor overload which accepts a `string8_view`.

Or perhaps the parameter could be a `source_code_span` instead of a `string8_view`?